### DISCUSSION TITLE
how are .msh files structured?

### DISCUSSION THREAD

**Discussion by poplarShift** [NOAA-EMC/WW3#628] [Q&A] [answered]
Hi all,

where can I find documentation on the format of the `*.msh` files for unstructured grids?

I've searched high and low through the repositories in different branches. From `regtests/ww3_tp2.7/input/ref1.msh` I've gathered that the list of nodes is index, lon, lat, z or something similar. 

But I'm used to the element list having having three columns (listing nodes of each cell), whereas the `.msh` files I've found have a variable amount of columns that even change mid-way. Unsure if this is related to the MeshFormat header but anyway, I'm confused. Can anybody enlighten me on this? 

Thanks!

**Comment by poplarShift** [comment]
For posteriority, I traced it back to a GMSH legacy format 2:

https://gmsh.info/doc/texinfo/gmsh.html#Legacy-formats

**Comment by aliabdolali** [comment, marked as answer]
Hi @poplarShift 
WW3 uses the GMSH format as you mentioned. Here is the simplified format, required by WW3
```
$MeshFormat
2 0 8
$EndMeshFormat
$Nodes
    <node number>
<node index> <longitude> <latitude> <depth>
$EndNodes
$Elements
        <element number>
open boundary nodes
land boundary nodes
   element index     2     3     0     1     5  <element conections (3)> 
$EndElements
```

The attached matlab script writes the WW3 *msh file if you have node, open boundary nodes and elements info.
[WW3_mesh_generator.m.zip](https://github.com/NOAA-EMC/WW3/files/8146670/WW3_mesh_generator.m.zip)
Here is the command you need,
**`WW3_mesh_generator(elelemt conection,longitude,latitude,depth,open boundary index,<name of the msh file>,0)`**

Regards
AA

**Reply by poplarShift** [reply]
Great, thank you so much!
